packages/tre/src/plugins/cache/gateway.ts

```diff
@@ -130,137 +142,186 @@ function getMatchResponse(
+        size += chunk.byteLength;
+        controller.enqueue(chunk);
+      },
+      flush() {
+        sizePromise.resolve(size);
+      },
+    });
+    this.size = sizePromise;
+  }
+}
+
 export class CacheGateway {
+  private readonly storage: KeyValueStorage<CacheMetadata>;
+
   constructor(
     private readonly log: Log,
-    private readonly storage: Storage,
+    legacyStorage: Storage,
     private readonly clock: Clock
-  ) {}
+  ) {
+    const storage = legacyStorage.getNewStorage();
+    this.storage = new KeyValueStorage(storage, clock);
+  }
 
   async match(request: Request, cacheKey?: string): Promise<Response> {
     // Never cache Workers Sites requests, so we always return on-disk files
     if (isSitesRequest(request)) throw new CacheMiss();
-
     cacheKey ??= request.url;
-    const cached = await this.storage.get<CacheMetadata>(cacheKey);
+
+    let resHeaders: Headers | undefined;
+    let resRanges: InclusiveRange[] | undefined;
+
+    const cached = await this.storage.get(cacheKey, ({ size, headers }) => {
+      resHeaders = new Headers(headers);
+      const contentType = resHeaders.get("Content-Type");
+
+      // Need size from metadata to parse `Range` header
+      const rangeHeader = request.headers.get("Range");
+      if (rangeHeader !== null) {
+        resRanges = _parseRanges(rangeHeader, size);
+        if (resRanges === undefined) throw new RangeNotSatisfiable(size);
+      }
+
+      return {
+        ranges: resRanges,
+        contentLength: size,
+        contentType: contentType ?? undefined,
+      };
+    });
     if (cached?.metadata === undefined) throw new CacheMiss();
 
-    const response = new CacheResponse(
-      cached.metadata,
-      cached.value
-    ).toResponse();
-    response.headers.set("CF-Cache-Status", "HIT");
+    // Avoid re-constructing headers if we already extracted multipart options
+    resHeaders ??= new Headers(cached.metadata.headers);
+    resHeaders.set("CF-Cache-Status", "HIT");
+    resRanges ??= [];
 
     return getMatchResponse(
       request.headers,
       cached.metadata.status,
-      response.headers,
-      cached.value
+      resHeaders,
+      resRanges,
+      cached.value,
+      cached.metadata.size
     );
   }
 
   async put(
     request: Request,
-    value: Uint8Array,
+    value: ReadableStream<Uint8Array>,
     cacheKey?: string
   ): Promise<Response> {
-    // Never cache Workers Sites requests, so we always return on-disk files
+    // Never cache Workers Sites requests, so we always return on-disk files.
     if (isSitesRequest(request)) return new Response(null, { status: 204 });
 
     const response = await HttpParser.get().parse(value);
+    let body = response.body;
+    assert(body !== null);
 
     const { storable, expiration, headers } = getExpiration(
       this.clock,
       request,
-      new Response(response.body, {
-        status: response.status,
-        headers: response.headers,
-      })
+      response
```

Comment:
I think this clone of the response may have been needed at some point... not sure though